### ISSUE TITLE
Cycle 4 Funding: Performance optimisation (Sinha)

### ISSUE THREAD

**Issue by jdswinbank** [astropy/astropy-project#413]
Issue to track progress of Cycle 4 funding request which was approved in https://github.com/astropy/astropy-project/pull/390 .

- [Project Description](https://github.com/astropy/astropy-project/blob/main/finance/proposal-calls/cycle4/manodeep-performance-optimisation-and-c-code-consultant.md)
- Awarded Amount: $10,200 (US)
- Period of Performance: 2024-05-01 through 2025-04-30
- Finance Committee Contact: @jdswinbank 
- COTR: @eteq

**Comment by manodeep** [comment]
Thanks John! 

I have created opencollective accounts both for my consulting biz and for myself - is there a preference for which one I should be submitting the invoices from? 

I am currently finishing up a couple of projects and so won't be able to begin until late-May/early-June. Presumably that also helps if selecting the COTR takes a bit longer than expected.

**Comment by manodeep** [comment]
@jdswinbank The period of performance has a typo - end date should (presumably) read as 2025-04-30 (and not *2024-04-30*)

**Comment by jdswinbank** [comment]
> I have created opencollective accounts both for my consulting biz and for myself - is there a preference for which one I should be submitting the invoices from?

Not from our side! I guess it might have tax implications or something for you, though...

> I am currently finishing up a couple of projects and so won't be able to begin until late-May/early-June. Presumably that also helps if selecting the COTR takes a bit longer than expected.

👍

**Comment by AnaGabela** [comment]
Hi, 
Let's get you started on getting a new contract with NumFocus because I think that's gonna take them a little while. Here are the instructions:

Go to https://numfocus.org/rocket
Click on Independent Contractor Registration
Form fields
Project: Astropy
Grant: NASA
Full Name
Email address
Street Address
Street Address
City
State
Postal Code
Country
Start Date: May 1, 2024-Feb 28, 2025
Description: Specreduce and spectroscopy development
Project Contact name: Astropy Finance Committee
Project Contact email: [finance@astropy.org](mailto:finance@astropy.org)
Payment method: ACH, Wire Transfer, PayPal
Bank Information
Comments about payment
Upload W9 or W8
Please let me know that you have submitted the form so we can keep track of it.

**Comment by manodeep** [comment]
Thanks @AnaGabela! The description looks like a copy-paste from a different proposal - do you want me to put "Performance optimisation" in the description?

**Comment by AnaGabela** [comment]
Oops... my mistake!
Yes please make the corrections!!

Thanks!

**Comment by manodeep** [comment]
Done!

**Comment by manodeep** [comment]
I have started working on the project. Here is the [viewable work log](https://hackmd.io/@manodeep/astropy-cycle4-perf-opt-log) that contains the time-log, the initial list of performance issues and my day-to-day work notes.

@jdswinbank @AnaGabela It would be good to have the COTR assigned - presumably I can talk to them about how to identify which issues/modules would benefit the most from a performance boost. (I can make some choices but those may not necessarily be of greatest community benefit)

**Comment by AnaGabela** [comment]
@manodeep yes we are looking for a COTR for this project. Can you give me some suggestions on who could take this role?
Thanks!

**Comment by manodeep** [comment]
Thanks, Ana! I don't know the constraints for a COTR - is there a list of potential COTR members? If it helps, I have worked with TomR and Brigitta (but I am unsure if either would be eligible + interested)

**Comment by AnaGabela** [comment]
Thanks Manodeep, we will discuss onMOnday with the finaance committee and let you know.

**Comment by AnaGabela** [comment]
Hi.. I just confirmed and @eteq will be the COTR for this project. Feel free to discuss how you will communicate.
Let me know if you have any questions.

**Comment by manodeep** [comment]
Thanks Ana!

**Comment by manodeep** [comment]
@AnaGabela I just submitted the invoice for my work done up to Aug. Please let me know if everything looks okay!

**Comment by AnaGabela** [comment]
@manodeep we just approved it and should all be good now.

**Comment by manodeep** [comment]
Fantastic - thanks!

**Comment by manodeep** [comment]
Update: Apologies for the delays at my end but because of life circumstances and starting a new full-time job, I have not been able to spend time on the project since ~Oct. Things should settle down in the next ~2 months and I can work on this again (during non-working hours).

If someone else wants to pick up on the work in the meantime, I am also happy to do a handover. The next steps that I had outlined was about adding function attributes to the C functions and compiling + testing - so should be straightforward for someone familiar with the build system.

**Comment by eteq** [comment]
Thanks for the update @manodeep !  I suspect no one else will have the bandwidth so you should continue as planned here when you find the time unless we explicitly hear that someone else is going to do it.

**Comment by jdswinbank** [comment]
Hey @manodeep ,

The Finance Committee suggests that we extend the period of performance of this work through 2026-02-28. Given your current situation, we don't suggest allocating any more funds at this point (but if you have a specific reason for making an additional request, we'd be happy to discuss it).

Assuming you are happy to take this no-cost extension, please make the request to NumFOCUS by completing [this form](https://forms.monday.com/forms/1bb44c839787d25bd6ee4acd456db414?r=use1). You can use the following information:

Your information:
* **Contractor Name**
* **Contractor Contact**
* **Contact Email address**
* **Full Street Address**

Contract details:
* **Contract Type:** Amendment
* **Project:** Astropy 
* **Project PI Contact:** @jdswinbank 
* **Project PI Email:** john@swinbank.org
* **Grant Type:** NASA
* **Grant Reference:** ROSES20 (20-OSTFL20-0057)
* **Hourly Rate:** $0
* **Contract Value:** $10,200
* **Start Date:** 2024-05-01
* **End Date:** 2026-02-28
* **Budget Category:** Development & Maintenance
* **Description of work:** Contractor will continue to work on Astropy performance enhancement and related activities.
* **Statement of Work:** Please use the text from your existing contract (which should be based on the [project description](https://github.com/astropy/astropy-project/blob/main/finance/proposal-calls/cycle4/manodeep-performance-optimisation-and-c-code-consultant.md)).
* **W9 (US workers) or W8 BEN (non-US workers)**

**Comment by manodeep** [comment]
@jdswinbank That's a great idea! I have submitted the contract amendment form.

**Comment by kelle** [comment]
From the finance committee's perspective, this project has become inactive. If this is incorrect or if you'd like to revisit this work, please re-open this issue or get in touch.

**Comment by manodeep** [comment]
Thanks @kelle - yes, this has certainly been inactive for a while but I would like to work on optimising astropy. The biggest barrier I faced was identifying what functions to optimise, and test-cases that serve as performance benchmarks. 

(Funnily enough, I emailed @eteq ~2 weeks ago asking to meet to discuss how to progress this project)
